### PR TITLE
T17565

### PR DESCRIPTION
**Pull request by wgreenberg**

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -7,6 +7,8 @@ const parseDataUrl = require('parse-data-url');
 const request = require('request');
 const url = require('url');
 const somaDOM = require('./soma_dom');
+const feedparser = require('feedparser');
+const moment = require('moment');
```
Comment (sgnn7) [inline lib/util.js]: Sort imports

File: lib/util.js [hunk 2]
```diff
@@ -193,3 +195,52 @@ function encode_uri(uri) {
...
+    const oldest_date = moment().subtract(max_days_old, 'days');
+
+    return _fetch_rss_page(feed_url, 1, [], max_items, oldest_date);
+}
+
+function _fetch_rss_json (feed_url) {
+    return new Promise((resolve, reject) => {
+        const req = request.get(feed_url);
+        const parser = new feedparser();
+        req.pipe(parser);
```
Comment (sgnn7) [inline lib/util.js]: Shouldn't this be after the error callback setting?

File: lib/util.js [hunk 3]
```diff
@@ -193,3 +195,52 @@ function encode_uri(uri) {
...
+// Assume a wordpress feed (https://codex.wordpress.org/Pagination#Adding_the_.22paged.22_parameter_to_a_query)
+function _fetch_rss_page (feed_url, page, items, max_items, oldest_date) {
+    let paginated_url = feed_url;
+    if (page > 1) {
+        paginated_url += `?paged=${page}`;
+    }
+
+    return _fetch_rss_json(paginated_url).then(feed_json => {
+        const recent_enough_items = feed_json.items.filter(article => article.pubdate >= oldest_date);
+        // If we've run into articles which are too old, cease crawling
```
Comment (sgnn7) [inline lib/util.js]: Is chronological ordering guaranteed?

File: lib/util.js [hunk 4]
```diff
@@ -193,3 +195,52 @@ function encode_uri(uri) {
...
+        if (recent_enough_items.length !== feed_json.items.length) {
+            return items.concat(recent_enough_items);
+        }
+
+        const new_items = items.concat(feed_json.items);
+        // If we've hit the desired number of items, also cease crawling
+        if (new_items.length >= max_items) {
+            return new_items.slice(max_items);
+        }
```
Comment (sgnn7) [inline lib/util.js]: Why not just concat the feed list directly with slice and then check the length which would prevent the need for `new_items` var completely?

File: lib/util.js (hunk 3 above)
Comment (wgreenberg) [inline lib/util.js]: well, it's guaranteed to be in the order of the feed

File: lib/util.js (hunk 2 above)
Comment (wgreenberg) [inline lib/util.js]: shouldn't matter, since events can't be emitted synchronously

File: lib/util.js (hunk 4 above)
Comment (wgreenberg) [inline lib/util.js]: could you give an example? not sure I get the suggestion

File: lib/util.js (hunk 3 above)
Comment (sgnn7) [inline lib/util.js]: Is that a wordpress thing or just something we've seen?

File: lib/util.js (hunk 4 above)
Comment (sgnn7) [inline lib/util.js]: Didn't try running this but this is what I was suggesting:
```
items.concat(feed_json.slice(0, items.length - max_items));
if (items.length == max_items) {
  return items;
}
```

Comment (wgreenberg) [inline lib/util.js]: concat is non-destructive, so we'd need a temporary variable anyhow -- IMO the more explicit (though potentially more verbose) way is clearer

Comment (sgnn7) [inline lib/util.js]: I get your point but it's a convoluted way of getting what you want - you add everything then if it's over the max then you slice out what you need vs getting up-to the exact amount you need and if it's the max returning. I don't feel strongly enough to hold up the PR over it but imo it's much more overhead for the marginal clarity it provides.

File: lib/util.js (hunk 2 above)
Comment (sgnn7) [inline lib/util.js]: hmm I would think that during execution of this block here the ops are still ordered so the `pipe` connecting would precede the handler for error which might not fire if something occurs between L209 and 211/212 executing. Maybe I'm missing something?

File: lib/util.js (hunk 4 above)
Comment (wgreenberg) [inline lib/util.js]: fair nuff, updated the code

File: lib/util.js (hunk 2 above)
Comment (wgreenberg) [inline lib/util.js]: ah turns out I misunderstood how the event loop handles this, I think you're right. Interestingly, if an event emitter ever emits an `'error'` event and there aren't any listeners, it'll just throw an exception

**Pull request merged**